File: arrays/maximumSubarray.py
from typing import List

class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
      # array with one element case
      if len(nums) == 1:
        return nums[0]

      subarraySum = maxSum = nums[0]
      
      for i in range(1, len(nums)):
        subarraySum = max(nums[i], subarraySum+nums[i])
        
        if subarraySum > maxSum:
          maxSum = subarraySum
      
      return maxSum


s = Solution()
arr = [-2,1,-3,4,-1,2,1,-5,4]
print(s.maxSubArray(arr))

File: arrays/maximumSubarray.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var maxSubArray = function (nums) {
  if (nums.length === 1) return nums[0];

  let sum = nums[0],
    maxSum = nums[0];

  for (let i = 0; i < nums.length; i++) {
    sum = Math.max(nums[i], sum + nums[i]);

    maxSum = Math.max(sum, maxSum);
  }
  return maxSum;
};

let nums = [-2, 1, -3, 4, -1, 2, 1, -5, 4];

let ans = maxSubArray(nums);
console.log(ans);
